### ISSUE TITLE
SSO with AzureAD fails using session credentials of AzureADjoined device

### ISSUE THREAD

**Issue by DjavanRoa** [derricksmith/phpsaml#27]
Hi,

Configured the plugin with AzureAD.
For the SSO scenario using Edge (Chromium) with the windows session profile on a AzureADjoined device, authentication fails with the following error on AzureAD page:
*AADSTS75011: Authentication method 'X509, MultiFactor' by which the user authenticated with the service doesn't match requested authentication method 'Password, ProtectedTransport'. Contact the GLPI (dev) application owner.*

If I open glpi a InPrivate window I get prompted login+password+MFA and login is successfull.

After reading AzureAD guidance (https://docs.microsoft.com/en-us/azure/active-directory/develop/single-sign-on-saml-protocol#requestauthncontext%E2%80%8B%E2%80%8B%E2%80%8B%E2%80%8B%E2%80%8B%E2%80%8B%E2%80%8B) and fiddling with the AuthnRequest I figured out that if I remove `Comparison="exact"` and change the AuthnContextClassRef to X509 the auth will pass.

I replaced
```xml
<samlp:RequestedAuthnContext Comparison="exact"> 
 <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport</saml:AuthnContextClassRef>
</samlp:RequestedAuthnContext>
```
with 
```xml
<samlp:RequestedAuthnContext>
  <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:X509</saml:AuthnContextClassRef>
</samlp:RequestedAuthnContext>
```

I'm still trying to figure out what to change in the code to support both cases :)

Regards,

**Comment by DjavanRoa** [comment]
I found lines from 66 in 'lib/php-saml/advanced-settings.php" that seem related to this:
```xml
// Authentication context.
        // Set to false and no AuthContext will be sent in the AuthNRequest,
        // Set true or don't present this parameter and you will get an AuthContext 'exact' 'urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport'
        // Set an array with the possible auth context values: array ('urn:oasis:names:tc:SAML:2.0:ac:classes:Password', 'urn:oasis:names:tc:SAML:2.0:ac:classes:X509'),
        'requestedAuthnContext' => false,

        // Allows the authn comparison parameter to be set, defaults to 'exact' if
        // the setting is not present.
        'requestedAuthnContextComparison' => 'exact',
```
But unfortunatly I haven't managed to make it honor what ever settings I enter.

If found the part that's supposed to render these settings starting line 110 in "lib\php-saml\src\Saml2\AuthnRequest.php".

I still have to figure out if this is something broken or if it's a matter of settings.

**Comment by derricksmith** [comment]
I believe this is same issue in https://github.com/derricksmith/phpsaml/issues/21.  I haven't had time to research fully but I think you are on the right track.  I believe we can change exact to a different value but not certain.

**Comment by DjavanRoa** [comment]
I've tried other comparaison values but had no luck. Accepted values are {better, minimum, maximum, exact} (see "authnContextComparisonType" at https://www.ibm.com/support/knowledgecenter/SSD28V_liberty/com.ibm.websphere.liberty.autogen.nd.doc/ae/rwlp_config_samlWebSso20.html)

I finally managed to get it to work with all scenarios with a temp dirty fix.
AzureAD guidance for SAML said that the whole `<samlp:RequestedAuthnContext>...</samlp:RequestedAuthnContext>` bloc is actually optionnal. So I commented it at line 118 in "lib\php-saml\src\Saml2\AuthnRequest.php" and both PasswordProtectedTransport and X509 are now working. I think this also might be the issue with #21.

Here is what my deserialized SAMLRequest looks like now :
```xml
<samlp:AuthnRequest
    xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
    xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
    ID="ONELOGIN_fbd2d6b41b1b7ad18657988c40c25f20fxxxxxxx"
    Version="2.0"

    IssueInstant="2020-11-13T21:57:35Z"
    Destination="https://login.microsoftonline.com/00000000-aaaa-bbbb-cccc-111111111111/saml2"
    ProtocolBinding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
    AssertionConsumerServiceURL="https://myserver.domain.com/plugins/phpsaml/front/acs.php">
    <saml:Issuer>https://myserver.domain.com/</saml:Issuer>
    <samlp:NameIDPolicy
        Format="urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified"
        AllowCreate="true" />

</samlp:AuthnRequest>
```
Don't know if removing this has any impact though.

**Comment by DjavanRoa** [comment]
I got to the bottom of it.
The default value in advanced_settings is 'requestedAuthnContext' => false` which is what I want as it prevents rendering the RequestedAuthnContext bloc in the SAMLRequest. 
The thing is that advanced_settings.php is not loaded by the plugin, this explains why I was struggling in the first place 🤦 
What I ended up by doing is appending this setting at the end of settings.php and it did the trick.

If ever I've submitted a PR with this.

```php
// append to end of lib\php-saml\settings.php

    // Security settings
    'security' => array (
        'requestedAuthnContext' => false
    )
```
